crates/node/src/compiler.rs

```diff
@@ -0,0 +1,144 @@
+use alloy_primitives::{hex, Bytes, B256};
+use core::panic;
+use reth_primitives::Bytecode;
+use reth_revm::{
+    handler::register::EvmHandler,
+    interpreter::{InterpreterAction, SharedMemory, EMPTY_SHARED_MEMORY},
+    Context as RevmContext, Database, Frame,
+};
+use revmc::{primitives::SpecId, EvmCompilerFn, EvmLlvmBackend, OptimizationLevel};
+use std::{
+    collections::HashMap,
+    sync::{
+        mpsc::{Receiver, Sender},
+        Arc, Mutex,
+    },
+    thread,
+};
+
+#[derive(Debug)]
+pub struct ExternalContext {
+    sender: Sender<(B256, Bytes)>,
+    // TODO: cache shouldn't be here (and should definitely not be wrapped in a mutex)
+    cache: Arc<Mutex<HashMap<B256, EvmCompilerFn>>>,
+}
+
+impl ExternalContext {
+    pub fn new(spec_id: SpecId) -> Self {
+        let cache = Arc::new(Mutex::new(HashMap::new()));
+        let (sender, receiver) = std::sync::mpsc::channel();
+
+        // TODO: graceful shutdown
+        thread::spawn({
+            let cache = cache.clone();
+
+            move || {
+                let context = Box::leak(Box::new(revmc::llvm::Context::create()));
+                // TODO: fail properly here.
+                let backend =
+                    EvmLlvmBackend::new(context, false, OptimizationLevel::Aggressive).unwrap();
+                let mut compiler = revmc::EvmCompiler::new(backend);
+
+                while let Ok((hash, code)) = receiver.recv() {
+                    // Do we have to allocate here? Not sure there's a better option
+                    let name = hex::encode(hash);
+                    dbg!("compiled", &name);
+
+                    let result =
+                        unsafe { compiler.jit(&name, &code, spec_id) }.expect("catastrophe");
```

Comment:
I'm a little confused as to the relationship between a _compiler_, a _module_, and a _contract_. 
Should I be compiling all the code with the same compiler?